File: systems-for-design-and-implementation/labs/lab3/LibraryServicesImpl/src/library/services/impl/LibraryServerImpl.scala
package library.services.impl

import java.util
import java.util.concurrent.{ConcurrentHashMap, ExecutorService, Executors}

import library.model.{Book, BookLoan, User}
import library.persistance.repository.{IBookRepository, IUserRepository, RepositoryException}
import library.services.{ILibraryClient, ILibraryServer, LibraryException}

class LibraryServerImpl(var userRepository: IUserRepository, var bookRepository: IBookRepository) extends ILibraryServer
{
    private var librarianID: Int = -1
    private var loggedClients: util.Map[Integer, ILibraryClient] = null

    this.loggedClients = new ConcurrentHashMap[Integer, ILibraryClient]

    @throws[LibraryException]
    def login(user: User, client: ILibraryClient): User =
    {
        val isValid: Boolean = this.userRepository.verifyUser(user)
        if (!isValid)
        {
            throw new LibraryException("Authentication failed")
        }

        val new_user: User = userRepository.getUserByUsername(user.getUsername)
        if (isLogged(new_user))
        {
            throw new LibraryException("User already logged in")
        }
        if (new_user.isLibrarian)
        {
            if (librarianID == -1)
            {
                librarianID = new_user.getId
            }
            else
            {
                throw new LibraryException("Another librarian is already logged in")
            }
        }
        loggedClients.put(new_user.getId, client)
        new_user
    }

    private def isLogged(user: User): Boolean =
    {
        this.loggedClients.get(user.getId) != null
    }

    @throws[LibraryException]
    def logout(user: User, client: ILibraryClient)
    {
        val localClient: ILibraryClient = loggedClients.remove(user.getId)
        if (localClient == null)
        {
            throw new LibraryException("User " + user.getId + " is not logged in.")
        }
        if (user.isLibrarian)
        {
            librarianID = -1
        }
    }

    @throws[LibraryException]
    def getAvailableBooks: util.List[Book] =
    {
        try
        {
            bookRepository.getAllAvailable
        }
        catch
        {
            case e: RepositoryException => throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    def getAllLoans: util.List[BookLoan] =
    {
        try
        {
            bookRepository.getAllLoans
        }
        catch
        {
            case e: RepositoryException =>
                throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    def getBorrowedBy(user: User): util.List[Book] =
    {
        try
        {
            bookRepository.getAllBorrowedBy(user)
        }
        catch
        {
            case e: RepositoryException =>
                throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    def searchByTitle(title: String): util.List[Book] =
    {
        try
        {
            bookRepository.searchByTitle(title)
        }
        catch
        {
            case e: RepositoryException =>
                throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    def borrowBook(book: Book, user: User)
    {
        try
        {
            bookRepository.borrowBook(book, user)
            notifyUsersUpdateBook(book)
            notifyLibrariansBorrowBook(book.getId, user.getId)
        }
        catch
        {
            case e: RepositoryException =>
                throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    private def notifyLibrariansBorrowBook(book_id: Int, user_id: Int)
    {
        println("notifyLibrariansBorrowBook that loan updated: " + book_id + " | " + user_id)
        val loan: BookLoan = bookRepository.getLoan(book_id, user_id)
        if (librarianID == -1)
        {
            println("No librarians logged in")
            return
        }
        val executor: ExecutorService = Executors.newFixedThreadPool(1)
        val client: ILibraryClient = loggedClients.get(librarianID)
        assert(client != null)
        executor.execute(new Runnable()
        {
            def run()
            {
                try
                {
                    println("Notifying librarian [" + librarianID + "] that loan updated.")
                    client.updateLoan(loan)
                }
                catch
                {
                    case e: LibraryException =>
                        println("Error notifying user of book loan " + e)
                }
            }
        })
    }

    @throws[LibraryException]
    def returnBook(book: Int, user: Int)
    {
        try
        {
            bookRepository.returnBook(book, user)
            notifyUsersUpdateBook(this.bookRepository.getBookByID(book))
        }
        catch
        {
            case e: RepositoryException =>
                throw new LibraryException(e.getMessage)
        }
    }

    @throws[LibraryException]
    private def notifyUsersUpdateBook(old_book: Book)
    {
        println("notifyUsersUpdateBook that book changed: " + old_book)
        val new_book: Book = bookRepository.getBookByID(old_book.getId)
        val executor: ExecutorService = Executors.newFixedThreadPool(loggedClients.size)
        import scala.collection.JavaConversions._
        for (pair <- loggedClients.entrySet)
        {
            val client: ILibraryClient = pair.getValue
            val user_id: Int = pair.getKey
            if (user_id != -1 && user_id != librarianID)
            {
                executor.execute(new Runnable()
                {
                    def run()
                    {
                        try
                        {
                            println("Notifying [" + user_id + "] that book updated.")
                            client.updateBook(new_book)
                        }
                        catch
                        {
                            case e: LibraryException =>
                                println("Error notifying user of book update " + e)
                        }
                    }
                })
            }
        }
    }
}

File: systems-for-design-and-implementation/labs/lab2/LibraryServicesImpl/src/library/services/impl/LibraryServerImpl.java
package library.services.impl;

import library.model.Book;
import library.model.BookLoan;
import library.model.User;
import library.persistance.repository.IBookRepository;
import library.persistance.repository.IUserRepository;
import library.persistance.repository.RepositoryException;
import library.services.ILibraryClient;
import library.services.ILibraryServer;
import library.services.LibraryException;

import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class LibraryServerImpl implements ILibraryServer
{
    private IUserRepository userRepository;
    private IBookRepository bookRepository;

    // unique <>_<>
    private int librarianID = -1;
    private Map<Integer, ILibraryClient> loggedClients;

    public LibraryServerImpl(IUserRepository userRepository, IBookRepository bookRepository)
    {
        this.userRepository = userRepository;
        this.bookRepository = bookRepository;
        this.loggedClients = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized User login(User user, ILibraryClient client) throws LibraryException
    {
        boolean isValid = userRepository.verifyUser(user);
        if (!isValid)
        {
            throw new LibraryException("Authentication failed");
        }

        // we get the full user
        User new_user = userRepository.getUserByUsername(user.getUsername());
        if (isLogged(new_user))
        {
            throw new LibraryException("User already logged in");
        }

        // Only one librarian
        if (new_user.isLibrarian())
        {
            if (librarianID == -1)
            {
                librarianID = new_user.getId();
            }
            else
            {
                throw new LibraryException("Another librarian is already logged in");
            }
        }

        loggedClients.put(new_user.getId(), client);

        return new_user;
    }

    private boolean isLogged(User user)
    {
        return loggedClients.get(user.getId()) != null;
    }

    @Override
    public synchronized void logout(User user, ILibraryClient client) throws LibraryException
    {
        ILibraryClient localClient = loggedClients.remove(user.getId());
        if (localClient == null)
        {
            throw new LibraryException("User " + user.getId() + " is not logged in.");
        }

        // reset librarian
        if (user.isLibrarian())
        {
            librarianID = -1;
        }
    }

    @Override
    public synchronized List<Book> getAvailableBooks() throws LibraryException
    {
        try
        {
            return bookRepository.getAllAvailable();
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    @Override
    public synchronized List<BookLoan> getAllLoans() throws LibraryException
    {
        try
        {
            return bookRepository.getAllLoans();
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    @Override
    public synchronized List<Book> getBorrowedBy(User user) throws LibraryException
    {
        try
        {
            return bookRepository.getAllBorrowedBy(user);
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    @Override
    public synchronized List<Book> searchByTitle(String title) throws LibraryException
    {
        try
        {
            return bookRepository.searchByTitle(title);
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    @Override
    public synchronized void borrowBook(Book book, User user) throws LibraryException
    {
        try
        {
            bookRepository.borrowBook(book, user);
            notifyUsersUpdateBook(book);
            notifyLibrariansBorrowBook(book.getId(), user.getId());
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    @Override
    public synchronized void returnBook(int book, int user) throws LibraryException
    {
        try
        {
            bookRepository.returnBook(book, user);
            notifyUsersUpdateBook(this.bookRepository.getBookByID(book));
        }
        catch (RepositoryException e)
        {
            throw new LibraryException(e.getMessage());
        }
    }

    private void notifyLibrariansBorrowBook(int book_id, int user_id) throws LibraryException
    {
        System.out.println("notifyLibrariansBorrowBook that loan updated: " + book_id + " | " + user_id);
        final BookLoan loan = bookRepository.getLoan(book_id, user_id);

        if (librarianID == -1)
        {
            System.out.println("No librarians logged in");
            return;
        }

        ExecutorService executor = Executors.newFixedThreadPool(1);
        final ILibraryClient client = loggedClients.get(librarianID);
        assert client != null;

        executor.execute(new Runnable()
        {
            @Override
            public void run()
            {
                try
                {
                    System.out.println("Notifying librarian [" + librarianID + "] that loan updated.");
                    client.updateLoan(loan);
                }
                catch (LibraryException e)
                {
                    System.out.println("Error notifying user of book loan " + e);
                }
            }
        });
    }

    private void notifyUsersUpdateBook(Book old_book) throws LibraryException
    {
        System.out.println("notifyUsersUpdateBook that book changed: " + old_book);
        final Book new_book = bookRepository.getBookByID(old_book.getId());

        ExecutorService executor = Executors.newFixedThreadPool(loggedClients.size());
        for (Map.Entry<Integer, ILibraryClient> pair : loggedClients.entrySet())
        {
            final ILibraryClient client = pair.getValue();
            final int user_id = pair.getKey();

            // only users
            if (user_id != -1 && user_id != librarianID)
            {
                executor.execute(new Runnable()
                {
                    @Override
                    public void run()
                    {
                        try
                        {
                            System.out.println("Notifying [" + user_id + "] that book updated.");
                            client.updateBook(new_book);
                        }
                        catch (LibraryException e)
                        {
                            System.out.println("Error notifying user of book update " + e);
                        }
                    }
                });
            }
        }
    }
}
